src/book/02-system/01-foundations/07-director.mdx

```diff
@@ -0,0 +1,231 @@
+title: Director
+description: Introduction to the Director and how to use it to build behaviours.
+slug: /system/foundations/director
+---
+
+The NUbots use the director system to control all their behaviour elements.
+It allows composable behaviours that can take advantage of transition rules to ensure that the robot performs the right task at the right time.
+
+The Director does this by providing a set of new DSL keywords to NUClear that describe what tasks to run, what conditions are required, and how those tasks build on other tasks.
+It then uses these keywords to make a graph that it will use to decide which tasks to execute in order to achive the goals of the modules.
+
+The two DSL components that are added are Providers and Tasks.
+
+## Providers
+
+Providers are implemented as NUClear reactions and live in a NUClear `on` statement.
+They are the components which "provide" the functionality needed to perform a task, either by doing it directly or by running a series of subtasks.
+They form the basis of the modules that implement behaviours.
+Each provider will provide for a specific message type in a specific situation.
+For example a provider which executed a walk task, the code might look like:
+
+```cpp
+on<Provides<Walk>>().then(...);
+```
+
+Provider are triggered whenever a task needs to be executed by the director, however they can also be triggered using other NUClear keywords.
+The director will ensure that any provider which is not currently active will not run if it is not active, even if the other triggers would make it run.
+For example if the walk engine module wants to update at 10Hz when it is running a provider may look like.
+
+```
+on<Provides<Walk>, Every<10, Per<std::chrono::seconds>>().then(...);
+```
+
+This provider would then run once every time the walk command was updated, as well as running at 10Hz while the provider is active.
+
+If you do want a provider to run even when it is inactive you can make them optional (e.g. `on<Optional<Provides<Walk>>>().then(...);`).
+Then this will still run although its task will be empty.
+Non active providers are unable to emit subtasks and will show an error if they attempt to.
+
+Note that for the director to recognise subtasks as coming from a specific provider, those tasks MUST be emitted from that specific reaction.
+Otherwise they will be considered root tasks and can cause all sorts of chaos.
+
+### Provider Groups
+
+When there are multiple providers which provide for a single task type they are considered a provider group.
+When a new task is emitted the director will determine which of the providers in a group to run based on the current state of the system as well as any extra DSL keywords that are on the providers decleration.
+The director may also update which provider in a group is running if a state update means that the currently active provider can no longer run.
+In this case it will attempt to reassign that task to a provider which is able to run it.
```

Comment:
If there is no provider at all, that's a bug in the code and the director will yell at you.
For example if you try to run walk, but don't have a walk module installed it will yell at you.

On the other hand, if there is just no module that can run because of a state, like kick wants to run but the state doesn't match, Then the director will try to do things that will allow the kick to run. Or if it can't find one, it'll just wait and hope and run whatever the next highest priority thing is.